File: zero-server/src/main/kotlin/com/zeroday/service/NotificationService.kt
package com.zeroday.service

import com.zeroday.model.Notification
import com.zeroday.model.NotificationType
import com.zeroday.model.Player
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Player inbox. Each player has a bounded notification list (newest at
 * the end) so the client can render an unread badge. When the cap is
 * reached the oldest read notification is dropped first; unread
 * notifications are never silently deleted.
 */
class NotificationService(val perPlayerCap: Int = 100) {
    private val mutex = Mutex()

    suspend fun push(
        player: Player,
        type: NotificationType,
        title: String,
        message: String,
        data: Map<String, String> = emptyMap()
    ): Notification = mutex.withLock {
        val n = Notification(type = type, title = title, message = message, data = data)
        player.notifications.add(n)
        trim(player)
        n
    }

    /**
     * Push a notification, but if an unread one of the same
     * [coalesceKey] was pushed within the last [windowMs] milliseconds,
     * *replace* it with a single merged notification whose data
     * `count` is the running total. This is what keeps the inbox from
     * being spammed when 12 achievements unlock at once or 5 world
     * events fire in the same second.
     *
     * Returns the (new or merged) notification.
     */
    suspend fun pushCoalescing(
        player: Player,
        type: NotificationType,
        title: String,
        message: String,
        coalesceKey: String,
        windowMs: Long = 5_000L,
        data: Map<String, String> = emptyMap()
    ): Notification = mutex.withLock {
        val now = System.currentTimeMillis()
        val existing = player.notifications.lastOrNull {
            !it.read &&
                it.type == type &&
                it.createdAt >= now - windowMs &&
                it.data["coalesceKey"] == coalesceKey
        }
        if (existing != null) {
            val prevCount = existing.data["count"]?.toIntOrNull() ?: 1
            val mergedData = existing.data.toMutableMap()
            mergedData["count"] = (prevCount + 1).toString()
            val updated = existing.copy(
                title = title,
                message = message,
                data = mergedData,
                createdAt = now
            )
            val idx = player.notifications.indexOf(existing)
            player.notifications[idx] = updated
            updated
        } else {
            val n = Notification(
                type = type,
                title = title,
                message = message,
                data = data + ("coalesceKey" to coalesceKey) + ("count" to "1")
            )
            player.notifications.add(n)
            trim(player)
            n
        }
    }

    suspend fun markRead(player: Player, notificationId: String): Boolean = mutex.withLock {
        val n = player.notifications.firstOrNull { it.id == notificationId } ?: return@withLock false
        n.read = true
        true
    }

    suspend fun markAllRead(player: Player): Int = mutex.withLock {
        var n = 0
        for (notif in player.notifications) {
            if (!notif.read) {
                notif.read = true
                n++
            }
        }
        n
    }

    fun unreadCount(player: Player): Int = player.notifications.count { !it.read }

    fun list(player: Player, includeRead: Boolean = true, limit: Int = 50): List<Notification> {
        val filtered = if (includeRead) player.notifications else player.notifications.filter { !it.read }
        return if (filtered.size <= limit) filtered else filtered.takeLast(limit)
    }

    private fun trim(player: Player) {
        if (player.notifications.size <= perPlayerCap) return
        val overflow = player.notifications.size - perPlayerCap
        // Drop oldest *read* notifications first to avoid losing unread state.
        val read = player.notifications.filter { it.read }
        val toRemove = read.take(overflow.coerceAtMost(read.size)).toSet()
        if (toRemove.isNotEmpty()) player.notifications.removeAll(toRemove)
        // If still over cap, drop oldest unread (cap is a hard ceiling).
        while (player.notifications.size > perPlayerCap) {
            player.notifications.removeAt(0)
        }
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Services/NotificationService.cs
using ZeroDayMMO.Shared.Models;

namespace ZeroDayMMO.Server.Services;

public interface INotificationService
{
    Notification AddNotification(string playerId, Notification notification);
    bool MarkRead(string playerId, string notificationId);
    int GetUnreadCount(string playerId);
    List<Notification> GetNotifications(string playerId, bool includeRead = true, int limit = 50);
}

public class NotificationService : INotificationService
{
    private readonly PlayerService _playerService;
    private readonly int _perPlayerCap;

    public NotificationService(PlayerService playerService, int perPlayerCap = 50)
    {
        _playerService = playerService;
        _perPlayerCap = perPlayerCap;
    }

    public Notification AddNotification(string playerId, Notification notification)
    {
        var player = _playerService.GetPlayer(playerId);
        if (player == null) throw new KeyNotFoundException("Player not found");

        lock (player.Notifications)
        {
            player.Notifications.Add(notification);
            Trim(player);
        }

        return notification;
    }

    public bool MarkRead(string playerId, string notificationId)
    {
        var player = _playerService.GetPlayer(playerId);
        if (player == null) return false;

        lock (player.Notifications)
        {
            var n = player.Notifications.FirstOrDefault(x => x.Id == notificationId);
            if (n == null) return false;
            n.Read = true;
            return true;
        }
    }

    public int GetUnreadCount(string playerId)
    {
        var player = _playerService.GetPlayer(playerId);
        if (player == null) return 0;

        lock (player.Notifications)
        {
            return player.Notifications.Count(n => !n.Read);
        }
    }

    public List<Notification> GetNotifications(string playerId, bool includeRead = true, int limit = 50)
    {
        var player = _playerService.GetPlayer(playerId);
        if (player == null) return new List<Notification>();

        lock (player.Notifications)
        {
            var filtered = includeRead
                ? player.Notifications.ToList()
                : player.Notifications.Where(n => !n.Read).ToList();

            if (filtered.Count <= limit) return filtered;
            return filtered.Skip(filtered.Count - limit).Take(limit).ToList();
        }
    }

    private void Trim(Player player)
    {
        if (player.Notifications.Count <= _perPlayerCap) return;

        var overflow = player.Notifications.Count - _perPlayerCap;
        var read = player.Notifications.Where(n => n.Read).ToList();
        var toRemove = read.Take(Math.Min(overflow, read.Count)).ToHashSet();

        if (toRemove.Count > 0)
            player.Notifications.RemoveAll(n => toRemove.Contains(n));

        while (player.Notifications.Count > _perPlayerCap)
        {
            player.Notifications.RemoveAt(0);
        }
    }
}
